### PR TITLE
Failing test case for resource acquired from another thread pool failing to clean up when stream is cancelled

### PR DESCRIPTION
**Pull request by jatcwang**
If you run the test you'll see "starting acq" and "acq" being printed out but not "released".

Without evalOn, the resource is successfully released. 

This is a minimized example from from https://github.com/typelevel/fs2/pull/2072

### CODE REVIEW

File: core/shared/src/test/scala/fs2/StreamSuite.scala [hunk 1]
```diff
@@ -550,6 +552,40 @@ class StreamSuite extends Fs2Suite {
...
+          println("starting acq")
+          Thread.sleep(1000)
+          println("acq")
+        }))(_ => IO(println("released")))
+
+      otherECRes.use { otherEC =>
+        (for {
+          st <- Ref.of[IO, List[String]](List.empty)
+          res <- IO.race(
+            Stream.resource(res(st, otherEC)).compile.toVector,
```
Comment (Daenyth) [inline core/shared/src/test/scala/fs2/StreamSuite.scala:578]: Does it also repro with this:
```suggestion
            res(st, otherEC).use(_ => IO.unit),
```

File: core/shared/src/test/scala/fs2/StreamSuite.scala [hunk 2]
```diff
@@ -550,6 +552,40 @@ class StreamSuite extends Fs2Suite {
...
+            IO.sleep(100.millis)
+          )
+          result <- st.get
+          _ <- IO.sleep(1.seconds)
```
Comment (mpilquist) [inline core/shared/src/test/scala/fs2/StreamSuite.scala:582]: Does this still fail if this sleep is 2 seconds?

Comment (jatcwang) [reply core/shared/src/test/scala/fs2/StreamSuite.scala:582]: Yes still fails with 5 sec

File: core/shared/src/test/scala/fs2/StreamSuite.scala (hunk 1 above)
Comment (jatcwang) [reply core/shared/src/test/scala/fs2/StreamSuite.scala:578]: Nope it's fine I get 

```
starting acq
acq
released
```

**Comment by mpilquist** [comment]
Fixes by #2073

**Comment by jatcwang** [comment]
Thanks for the fast fix @mpilquist, howeve it doesn't seem to fix the bug in doobie.

After minimizing, it seem like nested `Stream.resource` causes no release to occur. Here's the test case:

```
test(
      "Resource acquired on another thread pool is released when compiled stream is interrupted"
    ) {
      import java.util.concurrent.Executors
      import scala.concurrent.ExecutionContext
      import scala.concurrent.duration._
      val otherECRes =
        Resource.make(IO {
          scala.concurrent.ExecutionContext.fromExecutorService(Executors.newCachedThreadPool())
        })(ec =>
          IO {
            ec.shutdown
          }
        )
      def res(idx: Int, ec: ExecutionContext): Resource[IO, Unit] =
        Resource.make(munitContextShift.evalOn(ec)(IO {
          println(s"starting acq ${idx}")
          Thread.sleep(1000)
          println(s"acq ${idx}")
        }))(_ => IO(println(s"released ${idx}")))

      otherECRes.use { otherEC =>
        for {
          _ <- testCancelation(
            Stream
              .resource(res(0, otherEC))
              .flatMap(_ => Stream.resource(res(1, otherEC)))
          )
          _ <- IO.sleep(5.seconds) // give ample time for cleanups to happen
        } yield ()
      }
    }
```

For me it prints

```
starting acq 0
acq 0
```
(no cleanup happens)

Replacing the second Stream.resource with anything else seems to be fine.

**Comment by mpilquist** [comment]
@jatcwang Can you try again with 2.4-cecaf46?

**Comment by jatcwang** [comment]
That seem to have fixed it @mpilquist :)

**Pull request closed without merging**